src/services/refactors/convertToOptionalChainExpression.ts

```diff
@@ -0,0 +1,175 @@
+            actions: [{
+                name: refactorName,
+                description: convertToOptionalChainExpressionMessage
+            }]
+        }];
+    }
+
+    function getEditsForAction(context: RefactorContext, actionName: string): RefactorEditInfo | undefined {
+        const info = getInfo(context);
+        if (!info) return undefined;
+        const edits = textChanges.ChangeTracker.with(context, t => doChange(context.file, context.program.getTypeChecker(), t, info, actionName));
+        return { edits, renameFilename: undefined, renameLocation: undefined };
+    }
+
+    interface Info {
+        lastPropertyAccessChain: PropertyAccessExpression,
+        firstOccurrence: Node,
+        expression: BinaryExpression | ConditionalExpression
+    }
+
+    type ValidExpressionOrStatement = ValidExpression | ValidStatement;
+    type ValidExpression = BinaryExpression | ConditionalExpression;
+    type ValidStatement = ExpressionStatement | ReturnStatement | VariableStatement;
+
+    function isValidExpression(node: Node): node is ValidExpression {
+        return isBinaryExpression(node) || isConditionalExpression(node);
+    }
+
+    function isValidStatement(node: Node): node is ValidStatement {
+        return isExpressionStatement(node) || isReturnStatement(node) || isVariableStatement(node);
+    }
+
+    function isValidExpressionOrStatement(node: Node): node is ValidExpressionOrStatement {
+        return isValidExpression(node) || isValidStatement(node);
+    }
+
+    function getInfo(context: RefactorContext, considerEmptySpans = true): Info | undefined {
+        const { file, program } = context;
+        const span = getRefactorContextSpan(context);
+
+        const forEmptySpan = span.length === 0;
+        if (forEmptySpan && !considerEmptySpans) return undefined;
+
+        // selecting fo[|o && foo.ba|]r should be valid, so adjust span to fit start and end tokens
+        const startToken = getTokenAtPosition(file, span.start);
+        const endToken = getTokenAtPosition(file, span.start + span.length);
+        const adjustedSpan = createTextSpanFromNode(startToken, file, endToken);
+
+        const parent = forEmptySpan ? getParentForEmptySpanStartToken(startToken) : getParentNodeInSpan(startToken, file, adjustedSpan);
+        const expression = parent && isValidExpressionOrStatement(parent) ? getExpression(parent) : undefined;
+        if (!expression) return undefined;
+
+        const checker = program.getTypeChecker();
+
+        if (isBinaryExpression(expression)) {
+            const lastPropertyAccessChain = getLastPropertyAccessChain(expression.right);
+            if (!lastPropertyAccessChain) return undefined;
+            const firstOccurrence = getFirstOccurrence(expression, lastPropertyAccessChain, checker);
+            return firstOccurrence ? { lastPropertyAccessChain, firstOccurrence, expression } : undefined;
+        }
+
+        if (isConditionalExpression(expression)) {

```

Comment:
This makes it clearer that the cases are mutually exclusive:
```suggestion
        else if (isConditionalExpression(expression)) {
```